src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -235,6 +239,10 @@ public HubConnection(IConnectionFactory connectionFactory, IHubProtocol protocol
 
         _logScope = new ConnectionLogScope();
 
+        // ActivitySource can be resolved from the service provider when unit testing.
+        _activitySource = (serviceProvider.GetService<SignalRClientActivitySource>() ?? SignalRClientActivitySource.Instance).ActivitySource;
+        _serviceName = (_endPoint is UriEndPoint e) ? e.Uri.AbsolutePath.Trim('/') : null;
```

Comment:
I think this is fine, we don't know the class name on the client side and like you said elsewhere the server-side should have the route to match against already.

Do you know if other frameworks like gRPC have a similar problem?

Reply:
The client and server both know the service name with gRPC